backupccl: cache query results in data driven tests

msbutler opened the pull request:
Data driven test writers can now cache query results on a tag, and then compare
the query later in the test. Using this feature has a few benefits
- reduces the amount of text in a data driven test
- makes it a bit easier to reason about longer data driven test that conduct
  many comparisons
- opens the door for randomized testing in datadriven tests

A canonical example:
```
exec-sql tag=foo
SELECT * FROM foo;

[do some backup restore]

query-sql equals=foo
SELECT * FROM restoredFoo
```
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/84518)
<!-- Reviewable:end -->

adityamaru commented on pkg/ccl/backupccl/testdata/backup-restore/column-families:
```diff
@@ -46,16 +46,14 @@ exec-sql
 ----
 
-query-sql
+query-sql tag=rangeLen
```
hmm i'm not sure how I feel about eliding the output of a query. When reviewing and writing these tests the visual nature of the results being printed after the query and then confirming that what you (or the reviewer) sees is what you expect is a big part of ensuring the test works as expected IMO. It seems very easy for a test to remain broken silently *forever* if we tag, and just use that tag everywhere. wdyt?

adityamaru replied:
It also means the test reader/reviewer has to remember when the query was tagged, and then check that expecting the same values as that point in time is the correct behavior. I'm not too concerned about DD tests being verbose, checkout `spanconfigccl` for example. I don't have a good understanding about what you need for randomized testing but is tagging a pre-requisite?

msbutler replied:
hm yeah, these are all good points. The main application I see for this tool is when the test writer wants to verify that A == B, and doesn't really care what A is.

As an example, for my upcoming schema_only restore PR, I want to verify that the system descriptors are preserved before and after a cluster level schema_only restore, e.g. ensure that  `SELECT * FROM system.namespace WHERE parentID = 1` looks the same after the schema_only restore. I don't really care what the table looks like before the restore, and further, its a bit of an eyesore to have a ~50 line result in a data driven test.

Another near future application could be caching the results of `SELECT * FROM randomly_generated_table`. 

But again, your points are valid, and maybe I should add a big warning to the tool along the lines of : "only use this tool if you do not care about the contents of the cached query. You should only care that A == B"

shermanCRL replied:
What we are calling tags might more accurately be called variables. Maybe something like `save-sql var=foo`. Or `store-sql` or something.

adityamaru replied:
Or at the very least we should print it when we are tagging the output for the first time.

msbutler replied:
Aha, I like the idea of printing the output when you save the variable. I don't think this will work for randomized tests, but I don't see an immediate application for them yet, so we can worry about them later. 

@shermanCRL what about the flag `save-to-var=foo`? I.e.:

```
query-sql save-to-var=foo
SELECT * FROM foo;
----
aditya's
nice
printed
results

[do some backup restore]

query-sql equals-var=foo
SELECT * FROM restoredFoo
----
```

shermanCRL replied:
I like it.

msbutler committed:
backupccl: cache query results in data driven tests

Data driven test writers can now cache query results to a var, and then compare
the query later in the test. Using this feature has a few benefits

-reduces the amount of text in a data driven test
-makes it a bit easier to reason about longer data driven test that conduct
many comparisons

A canonical example:

exec-sql save-to-var=foo
SELECT * FROM foo;
----
results

[do some backup restore]

query-sql equals-var=foo
SELECT * FROM restoredFoo
----

Release note: None

adityamaru commented:
(trimming down my for review list)

msbutler commented:
no longer needed
